File: CPP-Version/src/UI/Window.cpp
//
// Created by steppers on 04/11/15.
//

#include "Window.h"
#include "../Engine/Managers/InputManager.h"

Window::Window(int width, int height) {
    _resolution = new Vector2i(width, height);
    _input = new Input();
    InputManager::getInstance().setInput(_input);
    init();
}

Window::~Window() {
    if(_validWindow)
    {
        glfwDestroyWindow(_window);
        glfwTerminate();
    }

    delete _resolution;
    delete _input;
}

void Window::init() {
    if (!glfwInit()){
        _validWindow = false;
        return;
    }

    _window = glfwCreateWindow(_resolution->getX(), _resolution->getY(), "-*|S|*-", NULL, NULL);
    if(!_window)
    {
        glfwTerminate();
        _validWindow = false;
        return;
    }
    glfwSetWindowUserPointer(_window, _input);  //Assign out input object to this window

    glfwMakeContextCurrent(_window);
    glfwSwapInterval(1);

    glfwSetKeyCallback(_window, key_callback);
    glfwMakeContextCurrent(nullptr);

    _validWindow = true;
}

void Window::getContext() {
    glfwMakeContextCurrent(_window);
}

void Window::releaseContext() {
    glfwMakeContextCurrent(nullptr);
}

void Window::update() {
    glfwSwapBuffers(_window);
}

bool Window::isCloseRequested() {
    return glfwWindowShouldClose(_window) == GL_TRUE;
}

bool Window::isValid() {
    return _validWindow;
}

Input* Window::getInput() {
    return _input;
}

File: src/UI/Window.java
package UI;

import Maths.Vector2i;
import org.lwjgl.opengl.GL;
import org.lwjgl.opengl.GL11;

import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.system.MemoryUtil.*;

/**
 * Created by steppers on 28/10/15.
 */
public class Window {

    private Vector2i resolution;
    private boolean fullscreen, vsync;
    private String title;

    private Input input;
    private long window;

    public Window(int width, int height, String title, boolean fullscreen, boolean vsync)
    {
        resolution = new Vector2i(width, height);
        this.fullscreen = fullscreen;
        this.vsync = vsync;
        this.title = title;
        init();
    }

    private void init()
    {
        if ( glfwInit() != GL11.GL_TRUE )
            throw new IllegalStateException("Unable to initialize GLFW");

        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
        glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);

        window = glfwCreateWindow(resolution.x, resolution.y, title, NULL, NULL);
        input = new Input(window);

        glfwMakeContextCurrent(window);
        glfwSwapInterval(1);
        glfwShowWindow(window);
    }

    public void update()
    {
        //glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        //glfwSwapBuffers(window);
        glfwPollEvents();
    }

    public boolean isCloseRequested()
    {
        return glfwWindowShouldClose(window) == GL_TRUE;
    }

    public void destroy()
    {
        input.release();
        glfwDestroyWindow(window);
        glfwTerminate();
    }

    public void getContext()
    {
        glfwMakeContextCurrent(window);
        GL.createCapabilities();
    }

    public void setResolution(int width, int height)
    {

    }

    public void setFullscreen(boolean fullscreen)
    {

    }

    public void setVsync(boolean vsync)
    {

    }

    public void setTitle(String title)
    {

    }

    public Input getInput()
    {
        return input;
    }

    public Vector2i getResolution()
    {
        return resolution;
    }

    public boolean isFullscreen()
    {
        return fullscreen;
    }

    public boolean vsyncEnabled()
    {
        return vsync;
    }

}
